container-search/src/main/java/com/yahoo/prelude/fastsearch/FS4InvokerFactory.java

```diff
@@ -43,11 +42,27 @@ public FS4InvokerFactory(FS4ResourcePool fs4ResourcePool, SearchCluster searchCl
     }
 
     public SearchInvoker getSearchInvoker(Query query, SearchCluster.Node node) {
-        return new FS4SearchInvoker(searcher, query, fs4ResourcePool, node);
+        Backend backend = fs4ResourcePool.getBackend(node.hostname(), node.fs4port());
+        return new FS4SearchInvoker(searcher, query, backend.openChannel(), node);
     }
 
     public Optional<SearchInvoker> getSearchInvoker(Query query, List<SearchCluster.Node> nodes) {
```

Comment:
Here you change the meaning of returning Optional.empty() from "don't dispatch internally" to "try another group" (I think). That looks right to me if you accounted for all uses of this method. It would be helpful with a javadoc comment which explained the semantics of returning empty (I think that's always a good rule - misunderstood empty() semantics is a common source of subtle bugs).

Reply:
Eventually the idea that Dispatcher can return an `Optional.empty` when `dispatchInternal` is active should go away entirely and the fallback would be an error saying "we don't have any backends available". Here with the invoker factory the change is as you say, no-value is now "this doesn't work" instead of "I don't know what to do". I'll add a javadoc before merging.